### PR TITLE
Purchasing Animals

### PR DESCRIPTION
**Pull request by dylbyl**
# Description
Users now have the ability to purchase an animal and add it to its corresponding facility. I have also added interfaces for various types of animals, mainly regarding what they produce. This will be helpful if we implement the Process Animal stretch goal.

## Related Tickets
1.  https://trello.com/c/sG2e31yM/17-users-can-purchase-a-grazing-animal
2. https://trello.com/c/Ou0EOqEs/18-users-can-purchase-chickens-or-ducks
3. https://trello.com/c/sPZckZdl/6-users-can-see-the-total-number-of-animals-in-a-selected-facility
4. https://trello.com/c/Vm0oiLha/5-users-can-only-choose-among-appropriate-facilities-for-a-given-animal

## Steps to Test:
- Save, add, and commit your own branch. Checkout and pull db-animals.
- Use dotnet run in the src folder to run the program.
- As a test, try choosing option 2 before we do any setup. Choose a cow. Make sure no facilities show up (none have been made yet!) and that you're prompted to press 0 to return to the main menu.
- Use option 1 to add a Grazing Field, Chicken House, and Duck House.
- Use option 2 to purchase an animal. For starters, choose a cow. (**Ticket 1**)
- After purchasing an animal, only appropriate facilities should show up. Add your animal to one. (**Ticket 4**)
- After adding a cow to a facility, try adding another to make sure that facility's animal count went up. (**Ticket 3**)
- Create at least 1 cow, 1 chicken, and 1 duck to ensure Grazing Fields, Chicken Houses, and Duck Houses all work. (**Ticket 2**)
- While adding animals, check that strings or numbers outside of the correct choices don't break the program.
- Try adding enough animals to fill a facility, and make sure that you can't add more beyond capacity.
- Use 0 at any step in the purchasing process to quit and return to the main menu.
- Once you're done adding, use option 4 on the main menu to pull up a report of the farm.

## What I added:
- Classes for every type of animal listed in the Readme.
- Interfaces for animals that eat feed or produce compost/feathers/eggs.
- Implemented each interface for every type of animal.
- Two new classes: ChooseChickenHouse and ChooseDuckHouse.
- New options and cases in PurchaseStock for each animal.
- Logic for options in the three Choose classes.
- Added a method to ChickenHouse and DuckHouse facilities that returns their current animal count.
- Changed ChickenHouse to be type IFacility<Chicken>, DuckHouse is now IFacility<Duck>.

# Checklist
- [x] When purchasing an animal, all 7 animal types are available to choose from. User can press 0 to return to main menu.
- [x] When choosing a grazing animal, user can choose a grazing field to add the animal to, or press 0 to return to main menu.
- [x] When choosing a chicken or duck, user can add to the appropriate type of house or press 0 to return to main menu.
- [x] If no facility exists for the animal to be placed in, users are prompted to press 0 to return to the main menu.
- [x] All facilities have counts beside them to show how many animals they currently have and what their full capacity is.
- [x] Users cannot add animals to a facility that is full. If a user tries, they are prompted again.
- [x] Users can only input numbers that correspond to an animal or facility choice. Entering wrong numbers or a string re-prompts them.
- [x] Adding facilities and then animals to those facilities are fully functional. They can be viewed with option 4 on the main menu.
- [x] My changes create no new errors.

### CODE REVIEW

**Comment by sarahebrooks12** [comment]
https://youtu.be/itKtosLoGk4

**Comment by DerekMayse** [comment]
Looks good: cant add animals to facilities that are at capacity (if you try you get  a helpful error message) but they are still shown as an option

**Review by barrygriffith84 — APPROVED** [review]
I can create all of the animals and facilities.
I can add animals to the correct facilities.
Only the correct facilities show up when I create an animal.
The number of animals in each facility prints correctly.
All of your error handling works.
I can successfully return to main menu by entering 0 on all of your menus.

Note:  Your code successfully prevents me from adding an animal to a facility that is full, but the facility still prints as an option.

Approved to merge.

File: src/Interfaces/ICompostAnimal.cs
Comment (jordan-castelloe) [inline src/Interfaces/ICompostAnimal.cs:4]: Are these interfaces necessary to complete the tickets on this PR? If not, they shouldn't be on the PR. Planning ahead is great, but coding ahead can get you into tight spots-- you shouldn't be writing code unless it specifically pertains to the ticket(s) you're working on

File: src/Interfaces/IFeeding.cs
Comment (jordan-castelloe) [inline src/Interfaces/IFeeding.cs:4]: For example: yes, you will eventually need interfaces for animals that produce compost, etc, in order to build the processing stretch goal. But where does this one come into play? Have you planned out the processing stretch goal completely, can you point to when and where this interface is necessary? If not, don't merge it. Excess interfaces will just make your code base more confusing and harder to get around.

File: src/Interfaces/IGrazing.cs
Comment (jordan-castelloe) [inline src/Interfaces/IGrazing.cs:8]: I think we already have `Type` property on our `IResource` interface-- why do we need both?

**Review by jordan-castelloe — COMMENTED** [review]
So in general, this works very well-- great job! I have one thing I'd like you to change before you merge:

1. [This ticket](https://trello.com/c/vpnBtqyG/10-user-can-only-choose-among-appropriate-facilities-that-have-enough-capacity) is not complete, at least with chickens. I can add 15 chickens to a chicken house and that chicken house will still show up on my list. It will prevent me from adding another chicken if it's at capacity, but it still shows it. I would recommend just tackling this ticket on another PR, and taking the link to this ticket and any associated testing instructions. You don't need to change anything about your code, but make sure you take it off the PR and move it back to the "to do" column.

In general, and when we test PRs in the future, we won't approve them if there's any code associated with tickets that you weren't working on. So the interfaces for `IComposting` etc, although technically correct, really shouldn't be on this PR. It's okay for right now, I won't make you redo them, but keep that in mind for next time-- coding too far ahead can often lead to more problems than it solves, so we're going to encourage you in the future to focus on the problem in front of you rather than trying to preemptively solve problems your team hasn't fully thought through yet.

File: src/Interfaces/IFeeding.cs
Comment (jordan-castelloe) [reply src/Interfaces/IFeeding.cs:4]: (you can merge it for now and delete it later if you end up not needing it, this is more for down the road)

**Pull request merged**
